controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -0,0 +1,349 @@
+        return streamMetadataStore.getCommittingTransactionsRecord(scope, stream, context, executor)
+                .thenCompose(record -> {
+                    if (record == null) {
+                        // no ongoing list transactions already chosen for commit.
+                        return createNewTxnCommitList(scope, stream, epoch, context, executor);
+                    } else {
+                        // check if the epoch in record matches current epoch. if not throw OperationNotAllowed
+                        if (record.getEpoch() == epoch) {
+                            // Note: If there are transactions that are not included in the commitList but have committing state,
+                            // we can be sure they will be completed through another event for this epoch.
+                            return CompletableFuture.completedFuture(record.getTransactionsToCommit());
+                        } else {
+                            log.debug("Postponing commit on epoch {} as transactions on different epoch {} are being committed for stream {}/{}",
+                                    epoch, record.getEpoch(), scope, stream);
+                            throw StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED,
+                                    "Transactions on different epoch are being committed");
+                        }
+                    }
+                });
+    }
+
+    /**
+     * 1. check idempotence --> check if all transactions are already committed, then rolling txn has already happened.
+     * 2. create duplicate txn epoch segments in segment store and commit transactions into those segments.
+     * 3. create duplicate active epoch segments
+     * 4. update history table with one complete and one partial epoch
+     * 5. seal active segments
+     * 6. complete partial record in history table
+     */
+    private CompletableFuture<Void> rollTransactions(String scope, String stream, HistoryRecord txnEpoch, HistoryRecord activeEpoch,
+                                                     List<UUID> transactionsToCommit, OperationContext context) {
+        // check if all transactions are already committed. if so return all good immediately
+        // just checking the last one suffices as we perform processing of transactions in order.
+        UUID lastTransactionId = transactionsToCommit.get(transactionsToCommit.size() - 1);
+        return streamMetadataStore.transactionStatus(scope, stream, lastTransactionId, context, executor)
+                .thenCompose(status -> {
+                    if (status.equals(TxnStatus.COMMITTING)) {
+                        return runRollingTxn(scope, stream, txnEpoch, activeEpoch, transactionsToCommit, context);
+                    } else {
+                        return CompletableFuture.completedFuture(null);
```

Comment:
I'm a bit confused by the logic here and the description at the top. I think we get here in the case that the txn status is `COMMITTED`, is it right?

Reply:
if last transaction in the list is already committed, then we have already completed the work and we can proceed to next phase (clean up of metadata barriers and state).